turbopack/crates/turbopack-core/src/chunk/mod.rs

```diff
@@ -108,6 +109,93 @@ impl ChunkableModules {
     }
 }
 
+/// A [Module] that can be merged with other [Module]s (to perform scope hoisting)
+// TODO currently this is only used for ecmascript modules, and with the current API cannot be used
+// with other module types (as a MergeableModule cannot prevent itself from being merged with other
+// module types)
+#[turbo_tasks::value_trait]
+pub trait MergeableModule: Module + Asset {
+    /// Even though MergeableModule is implemented, this allows a dynamic condition to determine
+    /// mergeability
+    #[turbo_tasks::function]
+    fn is_mergeable(self: Vc<Self>) -> Vc<bool> {
+        Vc::cell(true)
+    }
+
+    /// Create a new module representing the merged content of the given `modules`.
+    ///
+    /// Group entry points are not referenced by any other module in the group. This list is needed
+    /// because the merged module is created by recursively inlining modules when they are imported,
+    /// but this process has to start somewhere (= with these entry points).
+    #[turbo_tasks::function]
+    fn merge(
+        self: Vc<Self>,
+        modules: Vc<MergeableModulesExposed>,
+        entry_points: Vc<MergeableModules>,
+    ) -> Vc<Box<dyn ChunkableModule>>;
+}
+#[turbo_tasks::value(transparent)]
+pub struct MergeableModules(Vec<ResolvedVc<Box<dyn MergeableModule>>>);
+
+#[turbo_tasks::value_impl]
+impl MergeableModules {
+    #[turbo_tasks::function]
+    pub fn interned(modules: Vec<ResolvedVc<Box<dyn MergeableModule>>>) -> Vc<Self> {
+        Vc::cell(modules)
+    }
+}
+
+/// Whether a given module needs to be exposed (depending on how it is imported by other modules)
+#[derive(
+    Copy,
+    Clone,
+    Debug,
+    PartialEq,
+    Eq,
+    Serialize,
+    Deserialize,
+    TraceRawVcs,
+    NonLocalValue,
+    TaskInput,
+    Hash,
+)]
+pub enum MergeableModuleExposure {
+    // Only individual export of this module are used, and only within this group.
+    None,
+    // Individual exports and the namespace object of this module are used, but only within this
+    // group.
+    Internal,
```

Comment:
im confused about the difference between `None` and `Internal`.

is the difference about the `namespace` object being in use?  If so, shouldn't the `is_exposed` function just be `self == &Self::External`?  What is exposed in an `Internal`  exposure?

or is the issues that for `None`, no exports are used, even within the group?

Reply:
maybe MergableNamespaceExposure? since this isn't about the exposure of the 'module' but rather the need to expose the module namespace object?

or maybe this is fine, the docs help to clarify

Reply:
Internal vs External exists to be able to determine which modules have to be listed here: https://github.com/vercel/next.js/pull/79459#discussion_r2144389346